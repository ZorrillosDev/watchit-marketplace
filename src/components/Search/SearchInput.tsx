// REACT IMPORTS
import React, { FC, useState } from 'react';

// MUI IMPORTS
import { Box, BoxProps, Grid, GridProps, InputBase, InputBaseProps, styled } from '@mui/material';

// PROJECT IMPORTS
import { IconX } from '@tabler/icons';
import { Search as SearchIcon } from '@mui/icons-material';
import i18n from '@src/i18n';

/* eslint-disable  @typescript-eslint/strict-boolean-expressions */

// ===========================|| NOTIFICATIONS - ITEM ||=========================== //

export interface SearchInputProps {
  handleCancel: () => void
  onSearch: (txt: string) => void
}

const SearchInput: FC<SearchInputProps> = (props): JSX.Element => {
  const [text, setText] = useState('');

  const handleChange = (e: any): void => {
    setText(e.target.value);
    props.onSearch(e.target.value);
  };

  const handleCancel = (): void => {
    setText('');
    props.handleCancel();
  };

  return (
    <SearchInputWrapper container>
      <SearchIconWrapper filled={!!text}>
        <SearchIcon />
      </SearchIconWrapper>
      <SearchTextField
        placeholder={i18n.t('GLOBAL_SEARCH')}
        value={text}
        onFocus={() => props.onSearch(text)}
        onChange={handleChange}
      />
      {
        text
          ? (
            <SearchCancel onClick={handleCancel}>
              <IconX />
            </SearchCancel>
          )
          : <></>
      }
    </SearchInputWrapper>
  );
};

const SearchInputWrapper = styled(Grid)<GridProps>(() => ({
  borderRadius: '10px',
  flexWrap: 'nowrap',
  alignItems: 'center',
  justifyContent: 'flex-start',
  border: '1px solid rgb(229, 232, 235)',
  height: '2.5rem',
  width: '100%',
  zIndex: 1005,
}));

const SearchIconWrapper = styled(Box)<BoxProps & { filled: boolean }>(({ theme, filled }) => ({
  position: 'absolute',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  width: '2.5rem',
  height: '2.5rem',
  '& svg': {
    fill: filled ? theme.palette.primary.main : theme.palette.grey[400],
  },
}));

const SearchCancel = styled(Box)<BoxProps>(({ theme }) => ({
  position: 'absolute',
  cursor: 'pointer',
  right: 0,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  width: '2.5rem',
  height: '2.5rem',
  '& svg': {
    fill: theme.palette.primary.main,
    color: theme.palette.primary.main,
    width: '1.2rem',
    height: '1.2rem',
  },
}));

const SearchTextField = styled(InputBase)<InputBaseProps>(({ theme }) => ({
  marginTop: '2px',
  padding: 0,
  height: '100%',
  width: '100%',
  color: theme.palette.primary.main,
  '& input': {
    height: '100%',
    width: '100%',
    padding: '0 2.5rem 0 2.5rem !important',
    fontWeight: 'bold',
    color: theme.palette.primary.main,
  },
}));

export default SearchInput;
