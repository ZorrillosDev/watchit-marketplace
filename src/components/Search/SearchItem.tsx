// REACT IMPORTS
import React, { FC } from 'react';

// MUI IMPORTS
import { Grid, GridProps, styled, Typography } from '@mui/material';

// PROJECT IMPORTS
import { Color } from '@src/utils';
import { Movie } from '@state/movies/types';
import TruncatedTypography from '@components/TruncatedTypography';
import Poster from '@components/Poster';
import { IconStar } from '@tabler/icons';

// ===========================|| NOTIFICATIONS - ITEM ||=========================== //

const SearchItem: FC<Movie> = ({ ...movie }): JSX.Element => {
  return (
    <SearchItemWrapper item xs={12} display='flex' sx={{ p: '0' }}>
      <Poster image={movie.posters.small} name={movie.title} />
      <Grid container sx={{ p: 1 }}>
        <Grid item xs={9}>
          <Grid container>
            <Grid item xs={12}>
              <Typography variant='h4' color='primary'>
                {movie.title}
              </Typography>
            </Grid>
            <Grid item xs={12}>
              <TruncatedTypography lines={4} sx={{ m: 0 }}>
                {movie.synopsis}
              </TruncatedTypography>
            </Grid>
          </Grid>
        </Grid>
        <Grid item xs={3}>
          <Grid container>
            <Grid item xs={12}>
              <Typography variant='h5' color='secondary' textAlign='right' fontWeight='bold'>
                {movie.year}
              </Typography>
            </Grid>
            <Grid item xs={12}>
              <Typography variant='h5' color='secondary' textAlign='right'>
                <IconStar />
                {movie.rating}
              </Typography>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </SearchItemWrapper>
  );
};

const SearchItemWrapper = styled(Grid)<GridProps>(({ theme }) => ({
  cursor: 'pointer',
  padding: '16px',
  '& > .MuiCard-root': {
    width: '6rem',
  },
  '& svg': {
    display: 'inline',
    width: '1rem',
    height: '1rem',
    marginRight: '0.25rem',
    transform: 'translateY(3px)',
  },
  '&:hover': {
    background: Color.addAlpha(theme.palette.primary.light, 0.3),
  },
  '&:last-of-type': {
    borderBottomRightRadius: theme.shape.borderRadius,
    borderBottomLeftRadius: theme.shape.borderRadius,
  },
}));

export default SearchItem;
