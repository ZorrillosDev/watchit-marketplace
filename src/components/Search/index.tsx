// ===========================|| SEARCH EXPORTS ||=========================== //

export { Search as default } from '@components/Search/SearchContainer'
