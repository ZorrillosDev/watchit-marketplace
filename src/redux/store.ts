import { createStore, applyMiddleware, compose, Store } from 'redux'
import { routerMiddleware } from 'connected-react-router/immutable'
import thunk from 'redux-thunk'
import rootReducer from './reducers'
import { History } from 'history'

export default (history: History, initialState = {}): Store => {
  // devtools for debugging in dev environment.
  // @ts-expect-error
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?? compose

  // Array of all middlewares to be applied.
  const middlewares = [
    thunk,
    routerMiddleware(history)
  ]

  const enhancers = [
    applyMiddleware(...middlewares)
  ]

  return createStore(
    rootReducer(history),
    initialState,
    composeEnhancers(...enhancers)
  )
}
