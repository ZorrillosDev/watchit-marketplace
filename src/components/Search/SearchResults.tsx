// REACT IMPORTS
import React, { FC } from 'react'

// MUI IMPORTS
import {
  CardContent, Grid, Paper, Popper, Typography,
  styled, GridProps, CircularProgress
} from '@mui/material'

// THIRD PARTY IMPORTS
import { VirtualElement } from '@popperjs/core'

// PROJECT IMPORTS
import Transitions from '@components/Transitions'
import { MainCard } from '@components/Cards'
import scroller from '@styles/scroller'
import { Movie } from '@state/movies/types'
import { SxProps, Theme } from '@mui/system'
import SearchItem from '@components/Search/SearchItem'

/* eslint-disable  @typescript-eslint/strict-boolean-expressions */

// ===========================|| SEARCH - VIEW ||=========================== //

export interface SearchResultsProps {
  movies: Movie[]
  open: boolean
  searching: boolean
  anchorRef: VirtualElement | null
}

const SearchResults: FC<SearchResultsProps> = (props): JSX.Element => {
  return (
    <Popper
      placement='bottom-start'
      open={props.open}
      anchorEl={props?.anchorRef}
      role={undefined}
      transition
      id='search_results'
      disablePortal
      popperOptions={{
        modifiers: [
          {
            name: 'offset',
            options: {
              offset: [0, 5]
            }
          }
        ]
      }}
    >
      {({ TransitionProps }) => (
        <Transitions in={props.open} {...TransitionProps}>
          <Paper>
            <MainCard border={false} elevation={16} content={false} sx={SearchCardSx}>
              <CardContent
                sx={{
                  p: '16px 0 16px 10px !important',
                  display: 'flex',
                  alignItems: 'center',
                  justifyContent: 'center'
                }}
              >

                {
                  props.searching && <CircularProgress size={30} />
                }

                {
                  !props.searching
                    ? props.movies.length
                      ? (
                        <Grid container direction='column' spacing={1} sx={{ pl: 1 }}>
                          <SearchContent item xs={12}>
                            <Grid container direction='column' spacing={1}>
                              {
                                props.movies.map((movie, i) => <SearchItem movie={movie} key={i} />)
                              }
                            </Grid>
                          </SearchContent>
                        </Grid>
                        )
                      : (
                        <Typography textAlign='center' width={1}>
                          No items found
                        </Typography>
                        )
                    : <></>
                }
              </CardContent>
            </MainCard>
          </Paper>
        </Transitions>
      )}
    </Popper>
  )
}

const SearchCardSx: SxProps<Theme> = {
  borderRadius: '10px',
  boxShadow: 'rgba(0,0,0,0.16) 0px 4px 16px',
  '&:hover': {
    boxShadow: 'rgba(0,0,0,0.16) 0px 4px 16px'
  }
}

const SearchContent = styled(Grid)<GridProps>(({ theme }) => ({
  height: 1,
  maxHeight: 'calc(100vh - 205px)',
  overflowX: 'hidden',
  padding: '0px !important',
  ...scroller(theme)
}))

export default SearchResults
