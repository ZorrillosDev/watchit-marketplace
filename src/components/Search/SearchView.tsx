// REACT IMPORTS
import React, { FC } from 'react'

// MUI IMPORTS
import {
  Box, ClickAwayListener, CardContent, Grid, Paper,
  Popper, Typography, styled, GridProps, InputBase,
  InputBaseProps, CircularProgress, BoxProps
} from '@mui/material'

// PROJECT IMPORTS
import Transitions from '@components/Transitions'
import { MainCard } from '@components/Cards'
import i18n from '@src/i18n'
import scroller from '@styles/scroller'
import { Movie } from '@state/movies/types'
import { Search as SearchIcon } from '@mui/icons-material'
import { SxProps, Theme } from '@mui/system'
import SearchItem from '@components/Search/SearchItem'
import { IconX } from '@tabler/icons'

/* eslint-disable  @typescript-eslint/strict-boolean-expressions */

// ===========================|| SEARCH - VIEW ||=========================== //

export interface SearchViewProps {
  movies: Movie[]
  open: boolean
  searching: boolean
  handleClose: () => void
  handleCancel: () => void
  onSearch: (txt: string) => void
}

const SearchView: FC<SearchViewProps> = (props): JSX.Element => {
  const anchorRef = React.useRef(null)
  const [text, setText] = React.useState('')

  const handleChange = (e: any): void => {
    setText(e.target.value)
    props.onSearch(e.target.value)
  }

  const handleCancel = (): void => {
    setText('')
    props.handleCancel()
  }

  return (
    <>
      <ClickAwayListener onClickAway={props.handleClose}>
        {/* Search Input */}
        <SearchWrapper ref={anchorRef}>
          <SearchInputWrapper container>
            <SearchIconWrapper filled={!!text}>
              <SearchIcon />
            </SearchIconWrapper>
            <SearchInput
              placeholder={i18n.t('GLOBAL_SEARCH')}
              value={text}
              onFocus={() => props.onSearch(text)}
              onChange={handleChange}
            />
            {
              text
                ? (
                  <SearchCancel onClick={handleCancel}>
                    <IconX />
                  </SearchCancel>
                  )
                : <></>
            }
          </SearchInputWrapper>
          <Popper
            placement='bottom-start'
            open={props.open}
            anchorEl={anchorRef.current}
            role={undefined}
            transition
            id='search_results'
            disablePortal
            popperOptions={{
              modifiers: [
                {
                  name: 'offset',
                  options: {
                    offset: [0, 5]
                  }
                }
              ]
            }}
          >
            {({ TransitionProps }) => (
              <Transitions in={props.open} {...TransitionProps}>
                <Paper>
                  <MainCard border={false} elevation={16} content={false} sx={SearchCardSx}>
                    <CardContent
                      sx={{
                        p: '16px 0 16px 10px !important',
                        display: 'flex',
                        alignItems: 'center',
                        justifyContent: 'center'
                      }}
                    >
                      {
                        props.searching &&
                          <CircularProgress size={30} />
                      }
                      {
                        (!props.searching && props.movies.length)
                          ? (
                            <Grid container direction='column' spacing={1} sx={{ pl: 1 }}>
                              <SearchContent item xs={12}>
                                <Grid container direction='column' spacing={1}>
                                  {
                                    props.movies.map((movie, i) => <SearchItem movie={movie} key={i} />)
                                  }
                                </Grid>
                              </SearchContent>
                            </Grid>
                            )
                          : <></>
                      }
                      {
                        (!props.searching && !props.movies.length)
                          ? (
                            <Typography textAlign='center' width={1}>
                              No items found
                            </Typography>
                            )
                          : <></>
                      }
                    </CardContent>
                  </MainCard>
                </Paper>
              </Transitions>
            )}
          </Popper>
        </SearchWrapper>
      </ClickAwayListener>
    </>
  )
}

const SearchWrapper = styled(Grid)<GridProps>(() => ({
  flexShrink: 0,
  alignItems: 'center',
  justifyContent: 'center',
  position: 'relative',
  zIndex: 1000,
  '#search_results': {
    top: 'calc(100% + 2px) !important',
    left: '0 !important',
    width: '100%',
    transform: 'none !important',
    zIndex: 1010
  },
  '@media (max-width: 600px)': {
    width: 'auto',
    marginLeft: '1rem',
    marginRight: '0.5rem',
    flexGrow: 1,
    maxWidth: 'calc(100% - 7rem)'
  },
  '@media (min-width: 600px)': {
    width: '25%',
    marginRight: 0,
    flexGrow: 1
  },
  '@media (min-width: 960px)': {
    width: '30%',
    marginLeft: '2rem'
  }
}))

const SearchInputWrapper = styled(Grid)<GridProps>(() => ({
  borderRadius: '10px',
  flexWrap: 'nowrap',
  alignItems: 'center',
  justifyContent: 'flex-start',
  border: '1px solid rgb(229, 232, 235)',
  height: '2.5rem',
  width: '100%',
  zIndex: 1005
}))

const SearchCardSx: SxProps<Theme> = {
  borderRadius: '10px',
  boxShadow: 'rgba(0,0,0,0.16) 0px 4px 16px',
  '&:hover': {
    boxShadow: 'rgba(0,0,0,0.16) 0px 4px 16px'
  }
}

const SearchContent = styled(Grid)<GridProps>(({ theme }) => ({
  height: 1,
  maxHeight: 'calc(100vh - 205px)',
  overflowX: 'hidden',
  padding: '0px !important',
  ...scroller(theme)
}))

const SearchIconWrapper = styled(Box)<BoxProps & { filled: boolean }>(({ theme, filled }) => ({
  position: 'absolute',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  width: '2.5rem',
  height: '2.5rem',
  '& svg': {
    fill: filled ? theme.palette.primary.main : theme.palette.grey[400]
  }
}))

const SearchCancel = styled(Box)<BoxProps>(({ theme }) => ({
  position: 'absolute',
  cursor: 'pointer',
  right: 0,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  width: '2.5rem',
  height: '2.5rem',
  '& svg': {
    fill: theme.palette.primary.main,
    color: theme.palette.primary.main,
    width: '1.2rem',
    height: '1.2rem'
  }
}))

const SearchInput = styled(InputBase)<InputBaseProps>(({ theme }) => ({
  marginTop: '2px',
  padding: 0,
  height: '100%',
  width: '100%',
  color: theme.palette.primary.main,
  '& input': {
    height: '100%',
    width: '100%',
    padding: '0 2.5rem 0 2.5rem !important',
    fontWeight: 'bold',
    color: theme.palette.primary.main
  }
}))

export default SearchView
