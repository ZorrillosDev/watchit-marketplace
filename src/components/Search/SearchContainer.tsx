// TRANSFER MODAL IMPORTS
import React, { FC } from 'react'

// PROJECT IMPORTS
import SearchView from '@components/Search/SearchView'
import { Movie } from '@state/movies/types'

// ===========================|| SEARCH - CONTAINER ||=========================== //

/* eslint-disable  @typescript-eslint/strict-boolean-expressions */

export const SearchContainer: FC = (): JSX.Element => {
  const [searchMovies, setSearchMovies] = React.useState<Movie[]>([])
  const [searching, setSearching] = React.useState(false)
  const [open, setOpen] = React.useState(false)

  const handleClose = (): void => {
    setOpen(false)
  }

  const handleCancel = (): void => {
    setOpen(false)
    setSearching(false)
    setSearchMovies([])
  }

  const onSearch = (text: string): void => {
    if (!text) return setOpen(false)

    setSearching(true)
    setOpen(true)

    // TODO add movies here
    setTimeout(() => {
      setSearching(false)
      setSearchMovies([])
    }, 1000)
  }

  return (
    <SearchView
      {...{
        open,
        handleClose,
        onSearch,
        searching,
        handleCancel
      }} movies={searchMovies}
    />
  )
}
