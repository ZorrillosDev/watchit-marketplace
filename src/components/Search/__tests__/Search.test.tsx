import Search from '@components/Search'
import { shallow } from 'enzyme'
import React from 'react'
import { Provider } from 'react-redux'
import { store } from '@state/store'

describe('<Search />', () => {
  it('should render transfer modal', () => {
    const component = shallow(
      <Provider store={store}>
        <Search />
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })
})
