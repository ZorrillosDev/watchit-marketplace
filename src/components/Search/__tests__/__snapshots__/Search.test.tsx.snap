// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render transfer modal 1`] = `ShallowWrapper {}`;
